config/http_config.go

```diff
@@ -769,16 +861,20 @@ func (rt *oauth2RoundTripper) RoundTrip(req *http.Request) (*http.Response, erro
 		changed bool
 	)
 
-	if rt.clientSecret != nil {
+	// Fetch the secret if it's our first run or always if the secret can change.
+	if rt.clientSecret != nil && (rt.rt == nil || !rt.clientSecret.immutable()) {
 		var err error
-		secret, err = rt.clientSecret.fetch()
+		secret, err = rt.clientSecret.fetch(req.Context())
 		if err != nil {
 			return nil, fmt.Errorf("unable to read oauth2 client secret: %w", err)
 		}
+
+		if !rt.clientSecret.immutable() {
+			rt.mtx.RLock()
+			changed = secret != rt.lastSecret
+			rt.mtx.RUnlock()
+		}
 	}
-	rt.mtx.RLock()
-	changed = secret != rt.secret
-	rt.mtx.RUnlock()
 
 	if changed || rt.rt == nil {
```

Comment:
Would it be harmful if we kill this if (827 line) and do the changed = ... even on rt.rt == nil case? Bit simpler code and then we can remove the `|| rt.rt == nil` part from if 879 (also simpler). WDYT?

Reply:
Is this what you mean?